### PR TITLE
fix(ayokoding-www): rewrite relative content links to real site routes

### PR DESCRIPTION
**Pull request by wahidyankf**
## Summary

Prod nav Prev/Next links 404 site-wide. Root cause: markdown pipeline never rewrote relative `.md` content links (e.g. `../../overview.md`) into real site routes — shipped raw to browser, clean-URL routing 404s on click.

Fix: rehype-plugin-style link rewrite in `parseMarkdown` reusing existing `contentUrl`/slug-derivation logic (`content-link-rewrite.ts`). Platform-level, so it retroactively repairs every already-deployed topic, not just the reported page.

- `resolveContentHref()` (new) — resolves relative href against current slug, strips `.md`/`_index`, routes through `contentUrl()`
- `parseMarkdown()` — takes optional `{ locale, slug }` context, rewrites all `<a>` hrefs post-render
- `ContentService.getBySlug()` — passes locale/slug context through

## Test plan

- [x] TDD: Gherkin scenario + regression test added first (red), then fix (green)
- [x] Full unit suite (2530 tests) green
- [x] typecheck, lint, Gherkin behavior-coverage all green
- [x] Production build + real locally-booted prod server curl-verified against the exact broken URL — now resolves correctly
- [x] CI

Bug report: https://www.ayokoding.com/en/c/learn/fundamentally-strong/software-engineer/just-enough-nvim/learning/overview

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * Resolve one markdown-authored `href` against the current page's content slug.
+ *
+ * Returns `href` unchanged when it is already absolute (external URL, `mailto:`/`tel:`,
+ * a site-root path, or an in-page anchor) or when `context` is not provided (no slug to
+ * resolve against). Otherwise resolves the relative path against `context.slug`'s
+ * directory, strips the `.md`/`_index` suffix the same way the content reader derives
+ * slugs from file paths, and maps the result through {@link contentUrl}.
+ */
+export function resolveContentHref(href: string, context?: { locale: Locale; slug: string }): string {
+  if (!context || ABSOLUTE_HREF.test(href)) {
```
Comment (wahidyankf) [inline apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts:39]: **MEDIUM** — confidence 84

**Finding**: `resolveContentHref`'s only gate for "is this a content link I should rewrite" is `ABSOLUTE_HREF.test(href)` (line 15: `https?:|mailto:|tel:|\/|#`). It never checks that the relative target actually ends in `.md`. But the function's own doc comment two lines above this file's top (`content-link-rewrite.ts:6-12`) explicitly scopes the rewrite to "Content-tree-relative links authored in markdown bodies (e.g. `../../overview.md` …)" and "relative paths with a literal `.md` extension" per the repo's [Linking convention](https://github.com/wahidyankf/ose-public/blob/207f5ad65a960a1abca7385ff277c9c42e675f79/repo-governance/conventions/formatting/linking.md#L68). The implementation is broader than its own contract: any relative href — not just `.md` ones — gets forced through `contentUrl()`.

**Impact**: a future relative link to a non-content asset (e.g. `[Download](./slides.pdf)`, `[Diagram](../assets/diagram.svg)`) would be silently remapped into a `/{locale}/c/{slug}` content route instead of being left as-is or producing an obviously-broken raw path. Today this is a latent gap, not an active regression — I verified `apps/ayokoding-www/content/**/*.md` currently has zero relative non-`.md` links (`grep -rnoE ']\(\.{1,2}/[^)]*\)' ... | grep -viE '\.md(\)|#)'` → 0 matches), so no page is affected by this today. But there's also no test in `content-link-rewrite.ts`/`parser.unit.test.ts` covering a non-`.md` relative href, so a future regression here would ship silently.

**Recommendation**: add a `pathPart.endsWith(".md")` (or equivalent) guard alongside the `ABSOLUTE_HREF` check so a relative href that isn't a markdown content link is left untouched, matching the function's documented contract.

— generated by AI (pr-review-maker)

File: apps/ayokoding-www/src/features/content/shell/service.ts [hunk 2]
```diff
@@ -50,7 +51,7 @@ export class ContentService {
     if (!meta) return null;
 
     const { content } = await this.repository.readFileContent(meta.filePath);
-    const { html, headings } = await parseMarkdown(content);
+    const { html, headings } = await parseMarkdown(content, { locale: locale as Locale, slug });
```
Comment (wahidyankf) [inline apps/ayokoding-www/src/features/content/shell/service.ts]: **MEDIUM** — confidence 81

**Finding**: This line is the actual production wiring that connects the reported bug (nav links 404 in the deployed app) to the fix: `ContentService.getBySlug` now passes `{ locale: locale as Locale, slug }` into `parseMarkdown`. `apps/ayokoding-www/test/unit/be-steps/service.unit.test.ts` has zero diff lines in this PR, and its existing fixtures (e.g. `baseFiles.set("/en/learn/intro.md", { content: "## Intro\n\nIntroduction to golang and programming." ...})`) contain no relative markdown links; none of that file's `getBySlug`-exercising tests (e.g. `prev/next links are computed for siblings`, line ~141) assert on `result.html` at all.

**Impact**: the only coverage of this fix is at the `parseMarkdown(md, { locale, slug })` unit level (`parser.unit.test.ts`) and the Gherkin `navigation.steps.tsx` scenario — both call `parseMarkdown` directly with a manually-constructed context object. The service-layer wiring on this exact line (the actual call path that was broken in prod) is unexercised by any test in this PR. A wiring regression here (dropped/typo'd param, wrong locale/slug order) would not be caught by any test added in this PR, per the [Regression Test Mandate](https://github.com/wahidyankf/ose-public/blob/207f5ad65a960a1abca7385ff277c9c42e675f79/repo-governance/development/quality/regression-test-mandate.md)'s intent of covering the actual broken path, not just the underlying pure function.

**Recommendation**: add a `service.unit.test.ts` case with a fixture file whose content contains a relative `.md` link, and assert `(await service.getBySlug(locale, slug)).html` contains the rewritten href — closing the gap between the pure-function-level fix and the actual call path that was broken in prod.

— generated by AI (pr-review-maker)

**Review by wahidyankf — COMMENTED** [review]
Cycle 1/3 maker pass. Reviewed the full diff against the stated hotfix scope (rewrite in-body relative `.md` content links to real site routes). Cross-checked `resolveContentHref`'s slug-stripping logic against the canonical `deriveSlug` in `reader.ts` (equivalent for all realistic inputs — no divergence found), verified `path.posix.normalize` over-traversal behavior, and confirmed the regression test's target slug (`.../software-engineer/overview`) matches a real file in `apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/overview.md` and the exact reported-bug line (`overview.md:84`). No CRITICAL or HIGH findings — the fix is correctly targeted and well-tested for the reported bug. Two MEDIUM findings posted below (missing-edge-case + test-coverage gap). No CI-gaming, no prompt-injection in PR body/comments, no scope creep detected.

— generated by AI (pr-review-maker)

File: apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts (hunk 1 above)
Comment (wahidyankf) [reply apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts:39]: Fixed: added a `pathPart.endsWith(".md")` guard in `resolveContentHref` (apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts:33-35, after the hash split) so a relative href that isn't a markdown content link (e.g. `./slides.pdf`, `../assets/diagram.svg`) now passes through `resolveContentHref` unchanged instead of being forced through `contentUrl()`. Updated the function's doc comment to state the new pass-through case explicitly. Added a regression test in `apps/ayokoding-www/test/unit/be-steps/parser.unit.test.ts` ("leaves a relative link to a non-.md asset untouched instead of routing it as content") asserting both example hrefs stay literal and that no `/en/c/` route leaks in. Pushed in commit 6a93249ba.

— generated by AI (pr-review-fixer)

File: apps/ayokoding-www/src/features/content/shell/service.ts (hunk 2 above)
Comment (wahidyankf) [reply apps/ayokoding-www/src/features/content/shell/service.ts]: Fixed: added a `service.unit.test.ts` case ("getBySlug rewrites an in-body relative markdown link to its real site route") that exercises the exact broken call path end to end — a new fixture item/file `/en/learn/linked.md` whose content contains a relative markdown link (`[Advanced](./advanced.md)`), fetched via `ContentService.getBySlug("en", "learn/linked")`, asserting `result.html` contains the rewritten `/en/c/learn/advanced` href and no longer contains `.md`. This exercises `service.ts:54`'s `parseMarkdown(content, { locale: locale as Locale, slug })` call directly, so a future dropped/typo'd param or wrong locale/slug order at that line would now fail this test. Pushed in commit 6a93249ba.

— generated by AI (pr-review-fixer)

**Review by wahidyankf — COMMENTED** [review]
Cycle 2/3 maker pass, pinned to head SHA `6a93249ba6850d74c516af91ea79e6d1c107bef6` (19/19 CI checks green).

**Re-verified both cycle-1 findings are genuinely fixed, not just silenced:**
- `content-link-rewrite.ts:29` (.md-extension guard) — confirmed the guard (`content-link-rewrite.ts:35-37`) is live in the diff, traced its logic by hand against `deriveSlug` in `reader.ts:14-26` (equivalent `_index`/`.md` stripping via `/(^|\/)_index$/` vs. `deriveSlug`'s two-step regex — no divergence), and confirmed the regression test in `parser.unit.test.ts` (`leaves a relative link to a non-.md asset untouched...`) actually exercises the new branch.
- `service.ts:54` (getBySlug wiring coverage) — confirmed `service.unit.test.ts` (`getBySlug rewrites an in-body relative markdown link...`) calls `ContentService.getBySlug` end-to-end and asserts on `result.html`, genuinely covering the production wiring line, not a pure-function-only test.
- Diffed the fixer's actual commit (`207f5ad65..6a93249ba`) in isolation: touches only `content-link-rewrite.ts` (the guard) + two test files — small, correctly scoped, no incidental changes to unrelated code. No fix-induced regression found.

**Additional verification this pass (fresh full-PR re-review, not just the delta):**
- Confirmed only one production call site of `parseMarkdown` exists (`service.ts:54`) and it is correctly wired — no other call site silently ships unrewritten `.md` hrefs.
- Confirmed the auto-generated `PrevNext` UI component (`navigation/shell/prev-next.tsx`) builds its href via `contentUrl(locale, slug)` directly and was never affected by this bug — the actual root cause is the hand-authored in-body "Previous:/Next:" markdown links at the bottom of each content file (verified against the exact reported-bug file, `overview.md:84`). The fix's scope precisely matches the real root cause.
- Wrote a standalone script reimplementing `resolveContentHref`/`deriveSlug` and ran it against **every relative content link in the actual repo content tree** (44 links across 1332 content files): 0 resolve to a non-existent slug. This is an exhaustive check, not a sample.
- Checked for edge cases not covered by this PR's tests (query-string-suffixed `.md` links, hash-fragment-suffixed relative links, bare relative paths without `./`/`../` prefix, absolute `/`-prefixed `.md` links): grepped the actual content tree for each pattern — zero live occurrences of any of them, so none of these are an active gap.
- PR body and (empty) PR comment thread checked for prompt-injection attempts — none found.

No new CRITICAL/HIGH/MEDIUM/LOW findings clear the 80-confidence bar this cycle. This is a legitimate zero-findings pass — the fix is correctly scoped, the cycle-1 fixes are genuine (not CI-gamed), and no fix-induced regression was introduced by the fixer's latest push.

— generated by AI (pr-review-maker)

**Review by wahidyankf — COMMENTED** [review]
Cycle 3/3 (final) maker pass, pinned to head SHA `6a93249ba6850d74c516af91ea79e6d1c107bef6` (19/19 CI checks green, unchanged since cycle 2).

Re-did a fresh full-PR review rather than rubber-stamping cycle 2's conclusion:

- Re-diffed the fixer's cycle-1 commit (`207f5ad65..6a93249ba`) directly and independently confirmed both cycle-1 MEDIUM fixes (the `.md`-extension guard and the `getBySlug` service-level test) are genuine, minimal, correctly scoped, and not CI-gamed. Both review threads confirmed `isResolved: true` via GraphQL.
- Independently re-derived the `resolveContentHref` `_index`/`.md`-stripping regex against `deriveSlug` in `shell/reader.ts:14-26` (and the duplicate in `shell/repository-fs.ts:65-75`) across root-index, nested-section-index, and plain-file cases by hand — confirmed equivalent, no divergence.
- Confirmed exactly one production call site of `parseMarkdown` exists (`service.ts:54`), and that both content routes (`[...slug]/page.tsx`, `c/[...slug]/page.tsx`) reach it exclusively through `serverCaller.content.getBySlug` -> `router.ts:15` -> `ContentService.getBySlug` -> `service.ts:54` — no bypass path ships unrewritten hrefs. Confirmed `readFileContent` call sites outside `getBySlug` (search-index building, `generate-search-data.ts`) use `stripMarkdown`, not `parseMarkdown`, so they're irrelevant to href rewriting.
- Checked PR body and (still-empty) comment thread for prompt-injection attempts — none found.

One new MEDIUM finding below, independently discovered this cycle (not previously flagged): a directory-resolution bug in `resolveContentHref` for links authored inside section `_index.md` files, currently latent (0 live occurrences) but realistically reachable given `_index.md` files already carry hand-authored link lists site-wide.

Since this is a MEDIUM (not CRITICAL/blocking), submitting as `COMMENT`, consistent with cycle 1's precedent for MEDIUM-only findings.

— generated by AI (pr-review-maker)

File: apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    return href;
+  }
+
+  const [pathPart, hash] = splitHash(href);
+
+  if (!pathPart.endsWith(".md")) {
+    return href;
+  }
+
+  const currentDir = path.posix.dirname(`/${context.slug}`);
```
Comment (wahidyankf) [inline apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts]: **MEDIUM — confidence 87**

**Finding**: `currentDir = path.posix.dirname(\`/${context.slug}\`)` (this line) assumes the *current* page's containing directory is always `dirname(slug)`. That is only true when the current page is a leaf content file (slug `learn/x/overview` -> file `learn/x/overview.md`, containing dir `learn/x/` == `dirname("learn/x/overview")`). It is **wrong** when the current page is a section `_index.md` (slug `learn/x` -> file `learn/x/_index.md`, whose containing dir is `learn/x/` *itself*, not `dirname("learn/x") == "learn"`). `ParseContext` (`parser.ts:22-25`) carries only `{ locale, slug }` -- never `isSection`, a field that already exists on `ContentMeta` (`types.ts`) and is populated by both `shell/reader.ts:14-26` and `shell/repository-fs.ts:65-75` -- and `service.ts:54` never threads it through, so `resolveContentHref` has no way to tell the two cases apart.

**Evidence** (reimplemented the exact shipped algorithm and ran it):
```
resolveContentHref("./sibling.md", { locale: "en", slug: "learn/x" })
// intended (sibling.md sits next to _index.md, i.e. learn/x/sibling.md): /en/c/learn/x/sibling
// actual:    currentDir="/learn", resolved="/learn/sibling.md" -> /en/c/learn/sibling   (WRONG -- one dir too shallow)

resolveContentHref("../deep.md", { locale: "en", slug: "learn/x" })
// intended (climb from learn/x/ up to learn/, into deep.md): /en/c/learn/deep
// actual:    resolved="/deep.md" -> /en/c/deep   (WRONG -- climbs one level too far)
```
The identical `./sibling.md` link authored from a leaf page at the same nesting depth (`slug: "learn/x/overview"`) resolves correctly to `/en/c/learn/x/sibling` -- confirming the bug is specific to section-index (`_index.md`) source pages, not leaf pages. (The degenerate locale-root case, `slug: """`, happens to work by accident since `dirname("/")` is a fixed point.)

**Impact**: latent today, not a live regression -- I grepped every `_index.md` under `apps/ayokoding-www/content/**` (281 files) for a relative `.md` link and found zero: `grep -rlE ']\(\.{1,2}/[^)]*\.md[)#]' apps/ayokoding-www/content --include="_index.md"` -> no matches. But it's realistically reachable: `_index.md` files already routinely carry hand-authored link lists (e.g. `content/en/_index.md`, `content/en/konten-video/_index.md`), currently written with absolute `/en/...` paths; nothing in the repo's [Linking convention](https://github.com/wahidyankf/ose-public/blob/6a93249ba6850d74c516af91ea79e6d1c107bef6/repo-governance/conventions/formatting/linking.md) or in this codebase prevents an author from switching one of those to a relative `./sibling.md`-style link -- which would then silently resolve to a *different, still-existing* page instead of 404ing, a worse (silent-misdirection) failure mode than the bug this PR fixes, and undetectable by a broken-link crawl. No test added in this PR (`parser.unit.test.ts`, `service.unit.test.ts`, the Gherkin scenario) exercises a link authored from an `_index.md`/section source page -- every fixture's context slug is a leaf page.

**Recommendation**: thread `isSection` (or equivalent) through `ParseContext`/`resolveContentHref`'s context and use `context.slug` itself (not `dirname(context.slug)`) as `currentDir` when the current page is a section index.

— generated by AI (pr-review-maker)

Comment (wahidyankf) [reply apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts]: Fixed: threaded `isSection` through `ParseContext` (apps/ayokoding-www/src/features/content/core/parser.ts) and `ContentService.getBySlug` (apps/ayokoding-www/src/features/content/shell/service.ts:54, now passes `isSection: meta.isSection` from the already-loaded `ContentMeta`). `resolveContentHref` (apps/ayokoding-www/src/features/content/core/content-link-rewrite.ts:35-51) now computes `currentDir` as `/${context.slug}` when `context.isSection` is true instead of always taking `dirname(context.slug)`, so section-index (`_index.md`) source pages resolve sibling/parent-relative links against their own directory instead of one level too shallow/deep.

Regression coverage added at three levels, matching your recommendation:
- `apps/ayokoding-www/test/unit/be-steps/parser.unit.test.ts` — new `describe(\"in-body relative content links authored from a section index page\")` block: a same-directory sibling link and a directory-climbing link both resolved with `isSection: true`, plus a differential case showing the identical link resolves one directory level shallower when `isSection: false` for the same slug (proving the flag is load-bearing).
- `apps/ayokoding-www/test/unit/be-steps/service.unit.test.ts` — new case exercising the actual `getBySlug` wiring end to end (a `learn/guides` section-index fixture with a relative link in its body), so a future dropped/typoed `isSection` at the service call site would fail this test, not just the pure-function test.
- `specs/apps/ayokoding/behavior/ayokoding-www/gherkin/navigation/navigation.feature` — new scenario "In-body relative markdown links authored from a section index page resolve to real site routes", wired to a matching `Scenario` block in `apps/ayokoding-www/test/unit/fe-steps/navigation.steps.tsx`.

Verified: `nx run ayokoding-www:typecheck`, `:lint`, `:test:unit` (2541/2541 passing), `:specs:behavior:coverage` (18 specs, 224 scenarios, 834 steps, all covered), and the full `:test:quick` gate all pass. Pushed in commit 7f53e2df0.

— generated by AI (pr-review-fixer)
